enterprise/backend/src/metabase_enterprise/serialization/v2/extract.clj

```diff
@@ -62,21 +62,16 @@
     (eduction (map #(serdes/extract-all % extract-opts)) cat (model-set opts))))
 
 ;; TODO Properly support "continue" - it should be contagious. Eg. a Dashboard with an illegal Card gets excluded too.
-(defn- chase-for [targets func]
-  (loop [to-chase (set targets)
-         chased   #{}]
-    (let [[m i :as item] (first to-chase)
-          res            (func m i)
-          chased         (conj chased item)
-          to-chase       (set/union (disj to-chase item) (set/difference res chased))]
-      (if (empty? to-chase)
-        chased
-        (recur to-chase chased)))))
-
-(defn- gather-lineage [targets]
-  (set/union
-   (chase-for targets serdes/ascendants)
-   (chase-for targets serdes/descendants)))
+(defn- traverse-with [targets traverse-fn]
+  (loop [to-traverse (set targets)
+         traversed   #{}]
+    (let [[m i :as item] (first to-traverse)
```

Comment:
It's a shame we have to destructure like this, when this function would be otherwise so decoupled from serialization specifics. Shall we change `descendants` and `ascendants` to accept a 2-vec pair, so this could just be 
```clojure
(let [item (first to-traverse)
      res  (traverse-fn item)
```

Reply:
Hah, yeah, that would be nice. I even tried looking at clojure.walk as replacement, but it's too insistent on looking inside regular data structures.

I could change it and then use as `(traverse-with #(serdes/descendants (first %) (second %)))` to lower the churn. Still would be a quite universal function then.

Reply:
Yeah, that would be better I think.

Reply:
Now that I think about it, let's keep in here? This comment about `continue` is not giving me rest. 
:)

Reply:
What's that about anyway? I don't understand it